### ISSUE TITLE
[template/no-non-null-assertion] False positive when null assertion is required for type narrowing

### ISSUE THREAD

**Issue by KlementMultiverse** [angular-eslint/angular-eslint#2928]
## Problem

The `template/no-non-null-assertion` rule in `packages/eslint-plugin-template/src/rules/no-non-null-assertion.ts` flags null assertions that are genuinely necessary for type safety, making the rule too strict for real-world Angular templates.

## Reproduction

Consider an Angular template where a property can be `T | null` but you need to pass it to a component expecting `T`:

```html
<app-child [value]="data!.property"></app-child>
```

Where `data` is typed as `Data | null` in the component class. The assertion `data!` is required because:
- The parent component has already verified `data` is not null (e.g., through `*ngIf`)
- The child component's input binding requires a non-null value
- Removing the assertion breaks type safety

The rule currently flags this as unnecessary, even when the assertion is the only way to satisfy the type system without unsafe type casting or restructuring the template.

## Why this matters

This rule produces false positives that force developers to choose between:
1. Suppressing the rule with `eslint-disable` comments (defeating the rule's purpose)
2. Refactoring templates unnecessarily to avoid the assertion
3. Using unsafe type casts instead of explicit assertions

The rule should distinguish between truly unnecessary assertions and assertions that serve a real type-narrowing purpose.

## Expected behavior

The rule should:
- Allow null assertions when the target property requires a non-null type
- Allow assertions when the value's type in the template context is broader than the required type (e.g., `T | null` → `T`)
- Still flag unnecessary assertions on properties already typed as non-null

## Context

- Angular version: 15+
- ESLint plugin-template version: recent
- This affects any Angular app using strict null checks with template type checking enabled

---
*Contributed by [Klement Gunndu](https://github.com/KlementMultiverse)*

**Comment by KlementMultiverse** [comment]
Withdrawing this issue. Apologies for any inconvenience.
